### PR TITLE
audit: Port audit_checksum method to rubocop and add tests

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----
#569

### CODE REVIEW

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            audit_checksums(spec_body, name)
+            resource_blocks = find_all_blocks(spec_body, :resource)
+            resource_blocks.each do |rb|
+              _, _, resource_body = *rb
+              audit_checksums(resource_body, name, string_content(parameters(rb).first))
+            end
+          end
+        end
+
+        def audit_checksums(node, spec, resource = nil)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Maybe `resource_name`?

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+          if string_content(checksum).size != 64 && regex_match_group(checksum, /^\w*$/)
+            problem "#{msg_prefix}sha256 should be 64 characters"
+          end
+
+          unless regex_match_group(checksum, /^[a-fA-F0-9]+$/)
+            problem "#{msg_prefix}sha256 contains invalid characters"
+          end
+
+          problem "#{msg_prefix}sha256 should be lowercase" unless regex_match_group(checksum, /^[a-f0-9]+$/)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: If this line is >80 characters: use `unless ... do`

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          if string_content(checksum).size.zero?
+            problem "#{msg_prefix}sha256 is empty"
+            return
+          end
+
+          if string_content(checksum).size != 64 && regex_match_group(checksum, /^\w*$/)
+            problem "#{msg_prefix}sha256 should be 64 characters"
+          end
+
+          unless regex_match_group(checksum, /^[a-fA-F0-9]+$/)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Maybe make it a case-sensitive regex (`/i`) and skip`A-F`?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 4]
```diff
@@ -250,7 +256,7 @@ def caveats_strings
 
       # Returns the array of arguments of the method_node
       def parameters(method_node)
-        return unless method_node.send_type?
+        return unless method_node.send_type? || method_node.block_type?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Turn this into an `if` (or two); `unless .. ||` is hard to parse.

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 2 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Can I make it `return unless` and followed by `problem`?

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Sure!

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 3 above)
Comment (JCount) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Just to clarify for posterity, you do mean `case-insensitive` regex using `/i`, correct?

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Sorry: yes, thanks @JCount

**Comment by GauthamGoli** [comment]
@MikeMcQuaid a side point: if `sha256` gets deprecated / new hashes in future, then this would require refactoring. Should I refactor line 34 - 51 to not hardcode `sha256`?

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid a side point: if sha256 gets deprecated / new hashes in future, then this would require refactoring. Should I refactor line 34 - 51 to not hardcode sha256?
> Edit: May not be required as new hash would mean, just adding new conditions below line 51.

Nah, I think it's fine as-is; we'll be on sha256 for quite a while longer.

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Great work again, thanks!

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
+require_relative "./extend/formula_cop"
+
+module RuboCop
+  module Cop
+    module FormulaAudit
+      class Checksum < FormulaCop
+        def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          %w[Stable Devel HEAD].each do |name|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Let's just make these lowercase, I don't think `Devel` looks very nice.

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            next unless spec_node = find_block(body_node, name.downcase.to_sym)
+            _, _, spec_body = *spec_node
+            audit_checksums(spec_body, name)
+            resource_blocks = find_all_blocks(spec_body, :resource)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: We also need to get all resource blocks that aren't inside a stable/devel/head.

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+          end
+          if find_node_method_by_name(node, :md5)
+            problem "#{msg_prefix}MD5 checksums are deprecated, please use SHA256"
+          end
+
+          if find_node_method_by_name(node, :sha1)
+            problem "#{msg_prefix}SHA1 checksums are deprecated, please use SHA256"
+          end
+
+          checksum_node = find_node_method_by_name(node, :sha256)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Need to handle if this is `nil`; I hit this locally. Also, this doesn't seem to be constrained to the current `node` as far as I can tell.

File: Library/Homebrew/rubocops/checksum_cop.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+          if string_content(checksum).size != 64 && regex_match_group(checksum, /^\w*$/)
+            problem "#{msg_prefix}sha256 should be 64 characters"
+          end
+
+          unless regex_match_group(checksum, /^[a-f0-9]+$/i)
+            problem "#{msg_prefix}sha256 contains invalid characters"
+          end
+
+          return unless regex_match_group(checksum, /^[a-f0-9]+$/)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: `return unless regex_match_group(checksum, /[A-F]/)` seems correct for here.

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 6 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: This is not currently being done in `audit_specs`, right?

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 7 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Should an error be raised, if there is no `sha256` specified?

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 8 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Ah. This is embarrassing :/

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: We all make mistakes; it's embarrassing that I didn't spot it in review!

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 7 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: A RuboCop error: yes, I think so. I was seeing this on formulae that had them though so it may be a bug, too.

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 6 above)
Comment (JCount) [inline Library/Homebrew/rubocops/checksum_cop.rb]: I believe that the resource blocks are currently being audited as part of running `audit_specs`, but the code structure relies on the ResourceAuditor class, and the `audit_checksum` it contains. I think that gap might be part of the cause of this issue.

Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: I couldn't find any existing method/code in `audit.rb` which audits `resource` blocks that aren't inside stable/devel/head blocks, is this something we want to add?

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 7 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Do you mean, `find_node_method_by_name` returns `nil` even when there is `sha256` in node?

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 6 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Auditing the stable spec audits everything which is outside those blocks (as they are implicitly stable). The existing code is therefore fine, it's just this code that needs adjusted for that implicit stable resource handling. Make sense?

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 7 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/checksum_cop.rb]: I can't remember now helpfully but I think it's just that again we can have a stable/devel/head block that doesn't have a URL therefore doesn't need a checksum (although head urls don't need checksums either).

File: Library/Homebrew/rubocops/checksum_cop.rb (hunk 6 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/checksum_cop.rb]: Yeah. Got it. Thanks.
